File: ios/Classes/EventHandlers/CardDetailPayEventHandler.swift
//
//  SaleEventHandler.swift
//  edfapay_sdk
//
//  Created by Zohaib Kambrani on 03/03/2023.
//

import Foundation
import Flutter
import UIKit
import EdfaPgSdk 


class CardDetailPayEventHandler : NSObject, FlutterStreamHandler{
    
    var eventSink:FlutterEventSink? = nil
    
    private lazy var saleAdapter: EdfaPgSaleAdapter = {
        let adapter = EdfaPgAdapterFactory().createSale()
        adapter.delegate = self
        return adapter
    }()
    
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events

        if let params = arguments as? [String:Any],
           let order = params["EdfaPgSaleOrder"] as? [String : Any?],
           let payer =  params["EdfaPgPayer"] as? [String : Any?],
           let card =  params["EdfaPgCard"] as? [String : Any?]{
            
            
            let order = EdfaPgSaleOrder.from(dictionary: order)
            let payer = EdfaPgPayer.from(dictionary: payer)
            let card = EdfaPgCard.from(dictionary: card)
            
            
            let languageCode = (params["EdfaPayLanguage"] as? String) ?? "en"
            let language = languageFrom(code: languageCode)

            let recurring = (params["recurringInit"] as? Bool) ?? false

            
    
            // The precise way to present by sdk it self
            EdfaPayWithCardDetails(viewController: UIApplication.currentViewController()!)
                .set(order: order)
                .set(payer: payer)
                .set(card: card)
                .set(language: language)
                .set(recurring: recurring)
                .on(transactionFailure: { result, error in
                    debugPrint("native.transactionFailure.result ==> \(String(describing: result))")
                    debugPrint("native.transactionFailure.error ==> \(String(describing: error))")
                    
                    self.handleFailure(error: error ?? "Error")
                    
                })
                .on(transactionSuccess: { res, data in
                    debugPrint("native.transactionSuccess.response ==> \(String(describing: res))")
                    debugPrint("native.transactionSuccess.data ==> \(String(describing: data))")
                    
                    self.handleSuccess(response: data as! EdfaPgGetTransactionDetailsSuccess)

                }).initialize(
                    onError: { err in
                        self.handleFailure(error: err)

                    }
                )
            
        }
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
    
    private func onPresent(){
        debugPrint("onPresent :)")
        eventSink?(["onPresent" : ":)"])
    }
    
    private func handleSuccess(response: EdfaPgGetTransactionDetailsSuccess){
        debugPrint("native.transactionSuccess.data ==> \(String(describing: response.toJSON(root: "success")))")
        eventSink?(response.toJSON(root: "success"))
    }
    
    private func handleFailure(error:Any){
        
        
        if let e = error as? Array<String>{
            let error = [
                "result" : "ERROR",
                "error_code" : 100000,
                "error_message" : "\(error)",
                "errors" : e.map({ eMsg in
                    return [
                        "code" : 100000,
                        "message" : eMsg
                    ]
                }),
            ] as [String : Any]
            eventSink?(["error":error])
            return
        }
        
        if let e = error as? EdfaPgError{
            eventSink?(["error" : e.json()])
            return
        }
        
        if let e = error as? Encodable{
            eventSink?(e.toJSON(root: "failure"))
            return
        }
        
        let error = [
            "result" : "ERROR",
            "error_code" : 100000,
            "error_message" : "\(error)",
            "errors" : [
                "code" : 100000,
                "message" : "\(error)"
            ],
        ] as [String : Any]
        eventSink?(["error":error])
    }
    
}

extension CardDetailPayEventHandler : EdfaPgAdapterDelegate{
    
    func willSendRequest(_ request: EdfaPgDataRequest) {
        
    }
    
    func didReceiveResponse(_ reponse: EdfaPgDataResponse?) {
        
    }
    
}

File: android/src/main/kotlin/com/edfapg/flutter/sdk/eventhandlers/CardDetailPayEventHandler.kt
package com.edfapg.flutter.sdk.eventhandlers

import android.content.Context
import android.widget.Toast
import com.edfapg.flutter.sdk.helper.toMap
import com.edfapg.sdk.model.request.card.EdfaPgCard
import com.edfapg.sdk.model.request.order.EdfaPgSaleOrder
import com.edfapg.sdk.model.request.payer.EdfaPgPayer
import com.edfapg.sdk.model.response.base.error.EdfaPgError
import com.edfapg.sdk.model.response.gettransactiondetails.EdfaPgGetTransactionDetailsSuccess
import com.edfapg.sdk.toolbox.EdfaPayDesignType
import com.edfapg.sdk.toolbox.EdfaPayLanguage
import com.edfapg.sdk.views.edfacardpay.EdfaCardPay
import com.edfapg.sdk.views.edfacardpay.EdfaPayWithCardDetails
import com.google.gson.Gson
import io.flutter.plugin.common.EventChannel
import java.io.Serializable

class CardDetailPayEventHandler(private val context: Context): EventChannel.StreamHandler {
    var sink:EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        (arguments as? Map<*, *>)?.let {
            with(it) {
                (get("EdfaPgSaleOrder") as? Map<*, *>)?.let { orderMap ->
                    (get("EdfaPgPayer") as? Map<*, *>)?.let { payerMap ->
                        (get("EdfaPgCard") as? Map<*, *>)?.let { cardMap ->
                            val order = Gson().fromJson(Gson().toJson(orderMap), EdfaPgSaleOrder::class.java)
                            val payer = Gson().fromJson(Gson().toJson(payerMap), EdfaPgPayer::class.java)
                            val card = Gson().fromJson(Gson().toJson(cardMap), EdfaPgCard::class.java)
                            val locale = get("EdfaPayLanguage") as? String
                            val mLocale = EdfaPayLanguage.values().firstOrNull { it.value == locale} ?: EdfaPayLanguage.en
                            val recurring = (get("recurring") as? Boolean ?: false)
                            val auth = (get("auth") as? Boolean ?: false)

                            payWithCardDetails(order, payer, card, mLocale, recurring, auth)
                        }
                    }

                }
            }
        }
    }

    override fun onCancel(arguments: Any?) {

    }

    private fun payWithCardDetails(order:EdfaPgSaleOrder, payer:EdfaPgPayer, card:EdfaPgCard, language:EdfaPayLanguage, recurring:Boolean, auth:Boolean){
        EdfaPayWithCardDetails(context = context)
            .setOrder(order)
            .setPayer(payer)
            .setCard(card)
            .setRecurring(recurring)
            .setAuth(auth)
            .onTransactionFailure { res, data ->
                print("$res $data")
                handleFailure(data!!)

            }.onTransactionSuccess { res, data ->
                print("$res $data")
                handleSuccess(data as? EdfaPgGetTransactionDetailsSuccess)

            }.initialize(
                onError = {
                    handleFailure(it)
                },
                onPresent = {
                    onPresent()
                }
            )
    }


    private fun onPresent(){
        print("onPresent :)")
        sink?.success(
            mapOf(
                "onPresent" to ":)"
            )
        )
    }

    private fun handleSuccess(response: EdfaPgGetTransactionDetailsSuccess?){
        print("native.transactionSuccess.data ==> ${response?.toMap()}")
        sink?.success(
            mapOf(
                "success" to response?.toMap()
            )
        )
    }

    private fun handleFailure(error:Any){
        when (error) {
            is  EdfaPgError -> sink?.success(
                mapOf("error" to error.toMap())
            )
            is Serializable -> sink?.success(
                mapOf("failure" to error.toMap())
            )
            else -> sink?.success(
                mapOf(
                    "error" to mapOf(
                        "result" to "ERROR",
                        "error_code" to 100000,
                        "error_message" to "$error",
                        "errors" to listOf<Map<*,*>>(),
                    )
                )
            )

        }
    }
}